MDNetworking/MemberImpl/MDReplicatedMember.cs

```diff
@@ -0,0 +1,165 @@
+    {
+        MDLog.AddLogCategoryProperties(LOG_CAT, new MDLogProperties(MDLogLevel.Info));
+        this.Member = Member;
+        this.Reliable = Reliable;
+        this.ReplicatedType = ReplicatedType;
+        this.NodeRef = NodeRef;
+        CheckIfShouldReplicate();
+    }
+
+    ///<summary>Proviedes a unique key to identify this member</summary>
+    public String GetUniqueKey()
+    {
+        Node Node = NodeRef.GetRef() as Node;
+        return Node.GetPath() + "#" + Member.Name;
+    }
+
+    protected virtual object GetValue()
+    {
+        Node Instance = NodeRef.GetRef() as Node;
+        FieldInfo Field = Member as FieldInfo;
+
+        if (Field != null)
+        {
+            return Field.GetValue(Instance);
+        }
+
+        PropertyInfo Property = Member as PropertyInfo;
+        if (Property != null)
+        {
+            return Property.GetValue(Instance);
+        }
+
+        return null;
+    }
+
+    ///<summary>It is better to implement ReplicateToAll and ReplicateToPeer instead of this method</summary>
+    public virtual void Replicate(int JoinInProgressPeerId, bool IsIntervalReplicationTime)
+    {
+        object CurrentValue = GetValue();
+        Node Instance = NodeRef.GetRef() as Node;
+
+        if ((ReplicatedType == MDReplicatedType.Interval && IsIntervalReplicationTime) || 
+        (object.Equals(LastValue, CurrentValue) == false && ReplicatedType == MDReplicatedType.OnChange))
+        {
+            ReplicateToAll(Instance, CurrentValue);
+        }
+        else if (JoinInProgressPeerId != -1)
+        {
+            ReplicateToPeer(Instance, CurrentValue, JoinInProgressPeerId);
+        }
+    }
+
+    ///<summary>Replicate this value to all clients</summary>
+    protected virtual void ReplicateToAll(Node Node, object Value)
+    {
+        MDLog.Debug(LOG_CAT, "Replicating {0} with value {1} from {2}", Member.Name, Value, LastValue);
+        GD.Print("Replicating " + Member.Name + " from node " + Node.Name + " with value " + Value);
+        if (Reliable)
+        {
+            Node.Rset(Member.Name, Value);
+        }
+        else
+        {
+            Node.RsetUnreliable(Member.Name, Value);
+        }
+        LastValue = Value;
+    }
+
+    ///<summary>Replicate this value to the given peer</summary>
+    protected virtual void ReplicateToPeer(Node Node, object Value, int PeerId)
+    {
+        MDLog.Debug(LOG_CAT, "Replicating to JIP Peer {0} for member {1} with value {2}", PeerId, Member.Name, Value);
+        if (Reliable)
+        {
+            Node.RsetId(PeerId, Member.Name, Value);
+        }
+        else
+        {
+            Node.RsetUnreliableId(PeerId, Member.Name, Value);
+        }
+    }
+
+    public virtual bool ShouldReplicate()
+    {
+        // Note this means we don't support nodes swapping network master.
+        // If that should be supported then CheckIfShouldReplicate has to be run every update.
```

Comment:
I do have nodes that swap master in my game so I'll probably end up looking into changing this

Reply:
I will open a backlog item for this where we can look for a solution, I don't really like doing that entire check every frame since we can buffer it. Perhaps we need an this.ChangeNetworkMaster that we add as a node extension along with this.SpawnNetworkNode. Then we could at least hook that into the replicator and send an event so every other system can pick up that network master has been swapped.

Reply:
Backlog item has been opened, #38